### PR TITLE
feat: add audit log event

### PR DESCRIPTION
**Pull request by wardab0104**
feat: add audit log event

### CODE REVIEW

**Comment by Allen8975701** [comment]
@angelo-chan @marselab please help review

File: eventstream.go [hunk 1]
```diff
@@ -356,4 +359,262 @@ type Client interface {
...
+func NewAuditLogBuilder() *AuditLogBuilder {
+	return &AuditLogBuilder{
+		version:       defaultVersion,
+		ctx:           context.Background(),
+		errorCallback: nil,
+	}
+}
+
+type AuditLogPayload struct {
+	Content interface{}
```
Comment (Allen8975701) [inline eventstream.go:426]: json tag

Comment (Allen8975701) [reply eventstream.go:426]: map[string]interface{} ?

File: eventstream.go [hunk 2]
```diff
@@ -356,4 +359,262 @@ type Client interface {
...
+	Category        string           `json:"category,omitempty"`
+	ActionName      string           `json:"actionName,omitempty"`
+	Timestamp       int64            `json:"timestamp,omitempty"`
+	IP              string           `json:"ip,omitempty"`
+	Actor           string           `json:"actor,omitempty"`
+	ActorType       string           `json:"actorType,omitempty"`
+	ClientID        string           `json:"clientId,omitempty"`
+	ActorNamespace  string           `json:"actorNamespace,omitempty"`
+	ObjectIDs       []string         `json:"objectIds,omitempty"`
+	ObjectNamespace string           `json:"objectNamespace,omitempty"`
```
Comment (Allen8975701) [inline eventstream.go:376]: this should be required

File: eventstream.go [hunk 3]
```diff
@@ -356,4 +362,267 @@ type Client interface {
...
+	ID              string           `json:"id,omitempty"`
+	Category        string           `json:"category,omitempty"`
+	ActionName      string           `json:"actionName,omitempty"`
+	Timestamp       int64            `json:"timestamp,omitempty"`
+	IP              string           `json:"ip,omitempty"`
+	Actor           string           `json:"actor,omitempty"`
+	ActorType       string           `json:"actorType,omitempty"`
+	ClientID        string           `json:"clientId,omitempty"`
+	ActorNamespace  string           `json:"actorNamespace,omitempty"`
+	ObjectIDs       []string         `json:"objectIds,omitempty"`
```
Comment (angelo-chan) [inline eventstream.go:378]: here will we introduce the objectIds and targetUserIds, how 1:1 match the objectId with the targetUserId if there're multiple objects and users? 
or these 2 fields has no logic meaning, only for persistence and query?

File: eventstream.go [hunk 4]
```diff
@@ -356,4 +362,267 @@ type Client interface {
...
+	GetEventName() string
+	GetContext() context.Context
+	GetErrorCallback() errorCallbackFunc
+}
+
+type errorCallbackFunc func(event interface{}, err error)
+
+type AuditLogBuilder struct {
+	topics          []string
+	eventName       string
```
Comment (angelo-chan) [inline eventstream.go:398]: What's eventName mean?

File: eventstream.go [hunk 5]
```diff
@@ -356,4 +362,267 @@ type Client interface {
...
+	GetTopics() []string
+	GetEventName() string
+	GetContext() context.Context
+	GetErrorCallback() errorCallbackFunc
+}
+
+type errorCallbackFunc func(event interface{}, err error)
+
+type AuditLogBuilder struct {
+	topics          []string
```
Comment (angelo-chan) [inline eventstream.go:397]: does that mean we need to publish them on multiple topics? who will consume them?

File: eventstream.go [hunk 6]
```diff
@@ -356,4 +362,267 @@ type Client interface {
...
+	ClientID        string           `json:"clientId,omitempty"`
+	ActorNamespace  string           `json:"actorNamespace,omitempty"`
+	ObjectIDs       []string         `json:"objectIds,omitempty"`
+	ObjectNamespace string           `json:"objectNamespace"`
+	TargetUserIDs   []string         `json:"targetUserIds,omitempty"`
+	DeviceID        string           `json:"deviceId,omitempty"`
+	Payload         *AuditLogPayload `json:"payload,omitempty"`
+}
+
+type EventBuilder interface {
```
Comment (angelo-chan) [inline eventstream.go:385]: why EventBuilder and errorCallbackFunc? reuse?

File: eventstream.go [hunk 7]
```diff
@@ -356,4 +362,267 @@ type Client interface {
...
+		if topic == "" {
+			return errInvalidTopicEmpty
+		}
+
+		if valid := validateTopicEvent(topic); !valid {
+			return errInvalidTopicFormat
+		}
+	}
+
+	if auditLogBuilder.eventName == "" {
```
Comment (angelo-chan) [inline eventstream.go:532]: why events again?

File: kafka.go [hunk 8]
```diff
@@ -310,6 +310,74 @@ func (client *KafkaClient) publishEvent(ctx context.Context, topic, eventName st
 	return nil
 }
 
+// PublishAuditLog send an audit log event
+func (client *KafkaClient) PublishAuditLog(auditLogBuilder *AuditLogBuilder) error {
+	return client.PublishMessage(auditLogBuilder)
+}
+
+// PublishMessage send an event based on EventBuilder
+func (client *KafkaClient) PublishMessage(eventBuilder EventBuilder) error {
```
Comment (angelo-chan) [inline kafka.go:319]: I do not suggest mixing event with audit.

File: eventstream.go [hunk 9]
```diff
@@ -356,4 +362,267 @@ type Client interface {
 	Publish(publishBuilder *PublishBuilder) error
 	PublishSync(publishBuilder *PublishBuilder) error
 	Register(subscribeBuilder *SubscribeBuilder) error
+	PublishAuditLog(auditLogBuilder *AuditLogBuilder) error
+}
+
+type AuditLog struct {
```
[the remark this answers is not in the document]
Comment (wardab0104) [reply eventstream.go:368]: updated

File: validation.go [hunk 10]
```diff
@@ -36,6 +36,12 @@ var (
 	errInvalidSessionID       = errors.New("sessionID isn't valid")
 	errInvalidTraceID         = errors.New("traceID isn't valid")
 	errInvalidCallback        = errors.New("callback should not be nil")
+	errInvalidTopicEmpty      = errors.New("topic can not be empty")
```
Comment (Allen8975701) [inline validation.go:39]: where do we use these error？

**Review by marselsampe — APPROVED** [review]

**Review by Allen8975701 — APPROVED** [review]

File: README.md [hunk 11]
```diff
@@ -229,6 +229,51 @@ Event message format :
...
+
+#### Parameter
+* Category : Category. (string - Format: xxx_xxx - Required)
+* ActionName : Log action name. (string - Format: xxx_xxx - Required)
+* IP: IP address. (string - Optional)
+* Actor: Publisher/studio user id or client id. (string - UUID v4 without Hyphens - Required)
+* IsActorTypeUser: Is actor a user or not. (bool - true or false - Required)
+* ClientID : OAuth client ID. (string - UUID v4 without Hyphens - Required)
+* ActorNamespace : Actor namespace, should be publisher or studio namespace. (string - UUID v4 without Hyphens - Required)
+* ObjectID : Target resource id. (string - UUID v4 without Hyphens - Optional)
```
Comment (Allen8975701) [inline README.md:269]: please change description to match the actual validation rule

**Review by Allen8975701 — CHANGES_REQUESTED** [review]

File: kafka.go [hunk 12]
```diff
@@ -636,3 +685,50 @@ func (client *KafkaClient) runCallback(
 		Key:              event.Key,
 	}, nil)
 }
+
+func (client *KafkaClient) doPublish(ctx context.Context, topic string, config kafka.WriterConfig,
```
Comment (angelo-chan) [inline kafka.go:689]: this method seems similar to publishEvent, any reason we can't reuse it?

Comment (wardab0104) [reply kafka.go:689]: yes, I think we can reuse it

File: eventstream.go [hunk 13]
```diff
@@ -356,4 +362,197 @@ type Client interface {
...
+	auditLogBuilder.errorCallback = errCallback
+	return auditLogBuilder
+}
+
+func (auditLogBuilder *AuditLogBuilder) Build() (kafka.Message, error) {
+
+	id := generateID()
+	key := auditLogBuilder.key
+	if key == "" {
+		key = id
```
[the remark this answers is not in the document]
Comment (wardab0104) [reply eventstream.go:502]: yes, seems we don't need to change the key if it's empty, it's based on the balancer

**Review by angelo-chan — APPROVED** [review]

**Review by angelo-chan — APPROVED** [review]

**Pull request merged**